Fix Infinite processing of the same batch tag

sorin-iovita opened the pull request:
## Description

This PR fixes the infinite processing of the same batch tag id.
closes #970

Commit:
- introduce caching mechanism in federation batch processing to avoid infinite processing of the same batch tag
- externalize cache size

Commit:
Code refactor

Commit:
Code refactor

Commit:
Code refactor

hilmarf commented on services/download/src/main/java/app/coronawarn/server/services/download/FederationBatchProcessor.java:
```diff
@@ -25,7 +25,9 @@
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Optional;
+import java.util.Queue;
 import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.commons.collections4.queue.CircularFifoQueue;
```
in my opinion a bit over engineered... why isn't `java.util.Set` / HashSet sufficient?

roesslerj replied:
Removed the dependency and added a FIFO Queue with HashSet backing.

hilmarf commented on services/download/pom.xml:
```diff
@@ -20,6 +23,11 @@
       <artifactId>postgresql</artifactId>
       <scope>test</scope>
     </dependency>
+      <dependency>
+          <groupId>org.apache.commons</groupId>
+          <artifactId>commons-collections4</artifactId>
+          <version>${commons-collections4.version}</version>
+      </dependency>
```
I would like to not depend on more and more other libraries...

roesslerj replied:
Removed dependency.

hilmarf commented on services/download/src/main/resources/application.yaml:
```diff
@@ -19,6 +19,8 @@ services:
     validation:
       allowed-report-types: ${ALLOWED_REPORT_TYPES:CONFIRMED_TEST,CONFIRMED_CLINICAL_DIAGNOSIS}
+    # The number of batch tags that will be cached in memory
+    cached-batches-size: ${CACHED_BATCHES_SIZE:1000}
```
MAX_BATCH_COUNT

roesslerj replied:
Renamed it to "MAX_BATCH_LOOP_GUARD" as I think this better describes its purpose. But very open here :-)

roesslerj committed:
Use own simplistic implementation instead of importing additional library just for this

roesslerj committed:
Rename env variable CATCHED_BATCHES_SIZE to MAX_BATCH_LOOP_GUARD

hilmarf committed:
Merge branch 'master' into fix/denial-of-service-in-download-service

mibrasap commented on services/download/src/main/java/app/coronawarn/server/services/download/FederationBatchProcessor.java:
```diff
@@ -130,15 +133,21 @@ public void processUnprocessedFederationBatches() throws FatalFederationGatewayE
           .ifPresent(nextBatchTag -> {
-            if (config.getEfgsEnforceDateBasedDownload()) {
-              unprocessedBatches.add(new FederationBatchInfo(nextBatchTag, currentBatch.getDate()));
+            FederationBatchInfo batch = new FederationBatchInfo(nextBatchTag, currentBatch.getDate());
```
For clarity: please indicate in the variable name that it is only a batch info, but not the batch itself

roesslerj replied:
Done. Also in the line above (nextBatchTag).

mibrasap commented on services/download/src/main/java/app/coronawarn/server/services/download/FederationBatchProcessor.java:
```diff
@@ -42,6 +43,7 @@
   private final FederationGatewayDownloadService federationGatewayDownloadService;
   private final DownloadServiceConfig config;
   private final ValidFederationKeyFilter validFederationKeyFilter;
+  private final FifoMaxEntriesSet<FederationBatchInfo> cachedBatches;
```
Did you check if there is a simpler solution for that? What about a list that stores all the processed batch tags and compare incoming "next batch tags" against this list?

roesslerj replied:
Imho we need to limit the amount of batches we lookup, or else we're prone to a memory leak. Therefore a simple list doesn't suffice. But again—very open to different solution proposals :-)

roesslerj replied:
Will just use a HashSet.

mibrasap replied:
Nice, a set makes indeed more sense than a list in this case :)

mibrasap commented on services/download/src/test/java/app/coronawarn/server/services/download/FederationBatchProcessorTest.java:
```diff
@@ -226,6 +228,24 @@ void testOneUnprocessedBatchFails() throws Exception {
+      Mockito.verify(batchInfoService, times(1)).findByStatus(UNPROCESSED);
+      Mockito.verify(federationGatewayDownloadService, times(1)).downloadBatch(batchTag1, date);
+      Mockito.verify(batchInfoService, times(1)).updateStatus(batchInfo, PROCESSED);
+
```
unnecessary empty line

roesslerj replied:
Removed.

mibrasap commented on services/download/src/main/java/app/coronawarn/server/services/download/FederationBatchProcessor.java:
```diff
@@ -130,10 +134,14 @@ public void processUnprocessedFederationBatches() throws FatalFederationGatewayE
-              unprocessedBatches.add(new FederationBatchInfo(nextBatchTag, currentBatch.getDate()));
+              if (!cachedBatches.contains(batch)) {
+                unprocessedBatches.add(batch);
+              }
```
Do we really want to introduce a maximum number of batches that can be downloaded?

roesslerj replied:
I am also against this. I re-read the original ticket "This does not protect against a malicious server that sends different batch tags on each request." and I think we should have other means of guarding against a malicious servers in the backend (e.g. SSL etc.). We don't know how many cases there will be in Germany and Europe, so limiting the number of batches seems dangerous...

roesslerj replied:
As discussed: We'll run into OutOfMemory if there are too many batches, rather than setting a fixed limit...

roesslerj committed:
Minimalize and correct JavaDoc and implementation

hilmarf commented on services/download/src/main/java/app/coronawarn/server/services/download/FifoMaxEntriesSet.java:
```diff
@@ -0,0 +1,56 @@
+  public boolean add(E e) {
+    fast.add(e);
+    order.add(e);
+    while (fast.size() > maxSize) {
+      fast.remove(order.remove());
+    }
```
@sorin-stefan-iovita-sap / @roesslerj 
sorry, but I still don't get it... let's assume the following: we have max size of 3 and batch-IDs A, B, C, D, A, B, C, D, ...
when `fast` reaches `maxSize`, it contains A,B,C,D - it will remove A as it is the first element in `order`...
So with next batch being again ID: A, we are stuck in a loop again, aren't we?

roesslerj replied:
Yes. But there is a trade-off: Either no maxSize and possibly introduce memory leak—or have maxSize and not detect loops that are bigger than maxSize. I am ok with either, we just can't have both and need to decide what we want...

roesslerj replied:
Resolved as discussed :-)

roesslerj committed:
Simplified and corrected code as discussed:

We will eventually run into an OutOfMemoryException if there are too many batches. But that's better than setting a fixed limit and not detecting loops that are bigger than the limit OR only handle a fixed amount of batches altogether.

roesslerj committed:
Now we don't even need the env variable
